### PR TITLE
Helm templates for Ingress/Egress gateways + e2e v2 fixes

### PR DESCRIPTION
**Pull request by rshriram**
Adding (ineffective) helm templates for envoyv2 based gateways on ingress/egress.
Required for testing gateways
E2e v2 now passes with no RBAC, no auth and no ingress (i.e. http/grpc/tcp reachability, routing).

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  name: istio-ingress-gateway
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ template "istio.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    istio: ingress-gateway
+spec:
+  replicas: {{ .Values.replicaCount }}
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/ingress-gateway/templates/deployment.yaml]: .Values.ingressGateway.replicaCount or .Values.ingress.replicaCount.  I don't have enough context on the gateway functionality to understand if ingress and ingressGateway should go in separate namespaces in values.yaml.

File: install/kubernetes/helm/istio/requirements.yaml [hunk 2]
```diff
@@ -7,6 +7,14 @@ dependencies:
     version: 0.5.0
     # repository: file://../ingress
     condition: ingress.enabled
+  - name: ingress-gateway
```
Comment (sdake) [inline install/kubernetes/helm/istio/requirements.yaml]: I think there is a problem that helm does not allow -'s in the names.  I could be wrong.  Might be easier to just call this ingress_gateway or ingressgateway.  @jascott1 any thoughts?

File: install/kubernetes/helm/istio/values-istiotest.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@
 global:
   nodePort: true
   proxy:
-    image: proxy_debug
+    image: proxyv2
```
Comment (sdake) [inline install/kubernetes/helm/istio/values-istiotest.yaml]: this is for the testing deploy, so should be using proxyv2_debug if there is a debug container for proxyv2.  If not, looks good as is - with the idea to add a debug version of proxy.

File: install/kubernetes/helm/istio/values.yaml [hunk 4]
```diff
@@ -86,6 +86,40 @@ ingress:
       enabled: false
       port: 32000
 
+#
+# ingress-gateway configuration
+#
+ingress-gateway:
+  enabled: true
+  serviceAccountName: default
+  imagePullPolicy: IfNotPresent
```
Comment (sdake) [inline install/kubernetes/helm/istio/values.yaml]: it appears we have a global image pull policy in this helm chart - recommend just using that.

File: install/kubernetes/helm/istio/values.yaml [hunk 5]
```diff
@@ -86,6 +86,40 @@ ingress:
       enabled: false
       port: 32000
 
+#
+# ingress-gateway configuration
+#
+ingress-gateway:
+  enabled: true
+  serviceAccountName: default
```
Comment (sdake) [inline install/kubernetes/helm/istio/values.yaml]: I think, but am not certain if serviceAccountName is cruft that is left over.  I think we decided (but not certain) to hardcode service accounts.  @costinm would know more.

File: install/kubernetes/helm/istio/values.yaml [hunk 6]
```diff
@@ -86,6 +86,40 @@ ingress:
...
+#  cpu: 100m
+#  memory: 128Mi
+# requests:
+#  cpu: 100m
+#  memory: 128Mi
+
+#
+# egress-gateway configuration
+#
+egress-gateway:
```
Comment (sdake) [inline install/kubernetes/helm/istio/values.yaml]: same comments apply to egress-gateway as ingress-gateway

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.autoscaleMin }}
+apiVersion: autoscaling/v2beta1
+kind: HorizontalPodAutoscaler
+metadata:
+    name: istio-egress-gateway
+    namespace: {{ .Release.Namespace }}
+spec:
+    maxReplicas: {{ .Values.autoscaleMax }}
```
[the remark this answers is not in the document]
Comment (rshriram) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml]: I copied these verbatim frm ingress/autoscale.yaml

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.autoscaleMin }}
```
[the remark this answers is not in the document]
Comment (costinm) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml]: I think for each component we'll have min/max - it doesn't really make sense to have a single size.

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/clusterrole.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.global.rbacEnabled }}
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (costinm) [inline install/kubernetes/helm/istio/charts/egress-gateway/templates/clusterrole.yaml]: I know it's cut&pasted - but we should review if this is still needed. 
Gateways should not touch apiserver, just pilot

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  name: istio-egress-gateway
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ template "istio.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    istio: egress-gateway
+spec:
+  replicas: {{ .Values.replicaCount }}
```
[the remark this answers is not in the document]
Comment (costinm) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml]: Are you sure ? The other templates seem to use directly the values (ingress for example - I suspect this is a copy )

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/clusterrole.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.global.rbacEnabled }}
```
Comment (costinm) [inline install/kubernetes/helm/istio/charts/ingress-gateway/templates/clusterrole.yaml]: How about keeping the ingress as is (it is a gateway) ?

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (rshriram) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml]: yep. copied verbatim from ingress

[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml]: https://github.com/kubernetes/helm/blob/master/docs/chart_template_guide/subcharts_and_globals.md#overriding-values-from-a-parent-chart

```
There's an important detail to notice here. We didn't change the template of mychart/charts/mysubchart/templates/configmap.yaml to point to .Values.mysubchart.dessert. From that template's perspective, the value is still located at .Values.dessert. As the template engine passes values along, it sets the scope. So for the mysubchart templates, only values specifically for mysubchart will be available in .Values.

Sometimes, though, you do want certain values to be available to all of the templates. This is accomplished using global chart values.
```

ignore all my comments about scoping, unless you wish the value to be available to all of the templates.

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/autoscale.yaml]: agreed - when I left this comment it was at the start of the review and I had a misunderstanding about how variables are scoped in subcharts.  When I have worked on helm heavily in the past, there were not even subcharts as a thing with variable scoping :)

Cheers
-steve

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/autoscale.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.autoscaleMin }}
+apiVersion: autoscaling/v2beta1
+kind: HorizontalPodAutoscaler
+metadata:
+    name: istio-ingress-gateway
+    namespace: {{ .Release.Namespace }}
+spec:
+    maxReplicas: {{ .Values.autoscaleMax }}
+    minReplicas: {{ .Values.autoscaleMin }}
```
[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/ingress-gateway/templates/autoscale.yaml]: this is till a valid comment - values.yaml and the code don't match up.   I prefer autoscaleMin autoscaleMax.  The original variables appear to be autoScaleMin and autoAScaleMax.

File: install/kubernetes/helm/istio/values.yaml (hunk 4 above)
Comment (sdake) [reply install/kubernetes/helm/istio/values.yaml]: this comment was clear.  I was referring to this: https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/values.yaml#L19

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    metadata:
+      labels:
+        istio: egress-gateway
+      annotations:
+        sidecar.istio.io/inject: "false"
+    spec:
+      serviceAccountName: istio-egress-gateway-service-account
+      containers:
+        - name: {{ template "istio.name" . }}
+          image: "{{ .Values.global.hub }}/{{ .Values.global.proxy.image }}:{{ .Values.global.tag }}"
```
[the remark this answers is not in the document]
Comment (rshriram) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml]: to override for testing

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/service.yaml [hunk 14]
```diff
@@ -35,9 +35,10 @@ spec:
   type: NodePort
   ports:
   - port: 80
-    nodePort: {{ .Values.service.nodePort.port }}
+    nodePort: {{ .Values.service.nodePort.port80 }}
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/ingress-gateway/templates/service.yaml]: good catch, I missed this during review.

**Comment by sdake** [comment]
May not work for  the LB case but can be iterated upon if needed.

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/deployment.yaml]: right, thanks that makes sense now.  I think no tag needed for this scenario.

File: install/kubernetes/helm/istio/charts/egress-gateway/templates/clusterrole.yaml (hunk 9 above)
Comment (ZackButcher) [reply install/kubernetes/helm/istio/charts/egress-gateway/templates/clusterrole.yaml]: +1 - egress gateways don't need permissions, right?

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/clusterrole.yaml (hunk 11 above)
Comment (ZackButcher) [reply install/kubernetes/helm/istio/charts/ingress-gateway/templates/clusterrole.yaml]: Yup, we want this as a separate thing, but I don't think it should require any permissions (it should just talk to Pilot, right?).

File: tests/e2e/tests/pilot/istio_rbac_test.go [hunk 15]
```diff
@@ -28,9 +28,9 @@ const (
 )
 
 func TestRBAC(t *testing.T) {
-	if !tc.Kube.RBACEnabled {
-		t.Skipf("Skipping %s: rbac_enable=false", t.Name())
-	}
+	//if !tc.Kube.RBACEnabled {
```
Comment (ZackButcher) [inline tests/e2e/tests/pilot/istio_rbac_test.go]: Uncomment?

File: install/kubernetes/helm/istio/requirements.yaml (hunk 2 above)
Comment (rshriram) [reply install/kubernetes/helm/istio/requirements.yaml]: we have sidecar-injector

Comment (linsun) [reply install/kubernetes/helm/istio/requirements.yaml]: @sdake that is in values.yaml file we had to rename sidecar-injector to sidecarInjector.

File: install/kubernetes/helm/istio/values.yaml (hunk 6 above)
Comment (linsun) [reply install/kubernetes/helm/istio/values.yaml]: we had to change this to sidecarInjector in values.yaml.  it is probably good to change to that format to be consistent too.

File: install/kubernetes/helm/istio/requirements.yaml (hunk 2 above)
Comment (rshriram) [reply install/kubernetes/helm/istio/requirements.yaml]: It seems to work in the tests..

File: install/kubernetes/helm/istio/values-istiotest.yaml (hunk 3 above)
Comment (rshriram) [reply install/kubernetes/helm/istio/values-istiotest.yaml]: we don't have proxyv2_debug. Hence the choice

File: install/kubernetes/helm/istio/requirements.yaml (hunk 2 above)
Comment (linsun) [reply install/kubernetes/helm/istio/requirements.yaml]: see https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md.  really good work here, btw.

File: install/kubernetes/helm/istio/charts/ingress-gateway/templates/autoscale.yaml [hunk 16]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.autoscaleMin }}
```
[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/ingress-gateway/templates/autoscale.yaml]: ignore.

**Comment by rshriram** [comment]
@sdake its testing with whatever is in the ci/cd pipeline..

**Comment by sdake** [comment]
@rshriram cool - thats 2.7.2 then.

Thanks!
-steve

**Pull request merged**